bundle/ScheduledVisibility/Handler/ObjectState.php

```diff
@@ -0,0 +1,133 @@
+        private readonly int $visibleObjectStateId,
+        private readonly LoggerInterface $logger = new NullLogger(),
+    ) {}
+
+    public function hide(Content $content): void
+    {
+        try {
+            if ($this->isHidden($content)) {
+                return;
+            }
+        } catch (NotFoundException $e) {
+            $this->logger->error(
+                sprintf(
+                    'Configured object state group does not exist: %s',
+                    $e->getMessage(),
+                ),
+            );
+        }
+
+        $hiddenObjectStateId = $this->hiddenObjectStateId;
+
+        $this->toggleObjectState($content, $hiddenObjectStateId);
+    }
+
+    public function reveal(Content $content): void
+    {
+        try {
+            if (!$this->isHidden($content)) {
+                return;
+            }
+        } catch (NotFoundException $e) {
+            $this->logger->error(
+                sprintf(
+                    'Configured object state group does not exist: %s',
+                    $e->getMessage(),
+                ),
+            );
+        }
+
+        $visibleObjectStateId = $this->visibleObjectStateId;
+
+        $this->toggleObjectState($content, $visibleObjectStateId);
+    }
+
+    public function getType(): HandlerType
+    {
+        return HandlerType::ObjectState;
+    }
+
+    /**
+     * @throws NotFoundException
+     */
+    public function isHidden(Content $content): bool
+    {
+        $objectStateGroupId = $this->objectStateGroupId;
+        $objectStateGroup = $this->repository->sudo(
+            fn (): ObjectStateGroup => $this->objectStateService->loadObjectStateGroup($objectStateGroupId),
+        );
+
+        $objectState = $this->repository->sudo(
+            fn (): \Ibexa\Contracts\Core\Repository\Values\ObjectState\ObjectState => $this->objectStateService->getContentState($content->contentInfo, $objectStateGroup),
+        );
+
+        $hiddenObjectStateId = $this->hiddenObjectStateId;
+
+        return $hiddenObjectStateId === $objectState->id;
+    }
+
+    private function toggleObjectState(Content $content, int $objectStateId): void
+    {
+        try {
+            $objectState = $this->repository->sudo(
+                fn (): \Ibexa\Contracts\Core\Repository\Values\ObjectState\ObjectState => $this->objectStateService->loadObjectState($objectStateId),
+            );
+        } catch (NotFoundException $e) {
+            $this->logger->error(
+                sprintf(
+                    'Configured object state with id #%d does not exist: %s',
+                    $objectStateId,
+                    $e->getMessage(),
+                ),
+            );
+
+            return;
+        }
+
+        $objectStateGroupId = $this->objectStateGroupId;
+
+        try {
+            $objectStateGroup = $this->repository->sudo(
+                fn (): ObjectStateGroup => $this->objectStateService->loadObjectStateGroup($objectStateGroupId),
```

Comment:
`$objectState` loaded above already aggregates its `ObjectStateGroup`, so you can avoid loading the group separately.

Reply:
fixed in https://github.com/netgen/ibexa-scheduled-visibility/pull/1/commits/27ac190dc371bbccdba7b705d65ebdc9aa40bd30